src/modules/rc_update/rc_update.cpp

```diff
@@ -451,122 +424,200 @@ RCUpdate::Run()
-
-			/* copy from mapped manual_control_setpoint control to control group 3 */
-			actuator_controls_s actuator_group_3{};
-
-			actuator_group_3.timestamp = input_rc.timestamp_last_signal;
-
-			actuator_group_3.control[0] = manual_control_setpoint.y;
-			actuator_group_3.control[1] = manual_control_setpoint.x;
-			actuator_group_3.control[2] = manual_control_setpoint.r;
-			actuator_group_3.control[3] = manual_control_setpoint.z;
-			actuator_group_3.control[4] = manual_control_setpoint.flaps;
-			actuator_group_3.control[5] = manual_control_setpoint.aux1;
-			actuator_group_3.control[6] = manual_control_setpoint.aux2;
-			actuator_group_3.control[7] = manual_control_setpoint.aux3;
-
-			/* publish actuator_controls_3 topic */
-			_actuator_group_3_pub.publish(actuator_group_3);
+			// limit processing if there's no update
+			if (rc_updated || (hrt_elapsed_time(&_last_manual_control_setpoint_publish) > 300_ms)) {
+				UpdateManualSetpoint(input_rc.timestamp_last_signal);
+			}
+
+			UpdateManualSwitches(input_rc.timestamp_last_signal);
 
 			/* Update parameters from RC Channels (tuning with RC) if activated */
 			if (hrt_elapsed_time(&_last_rc_to_param_map_time) > 1_s) {
 				set_params_from_rc();
 				_last_rc_to_param_map_time = hrt_absolute_time();
 			}
 		}
+
+		memcpy(_rc_values_previous, input_rc.values, sizeof(input_rc.values[0]) * input_rc.channel_count);
+		static_assert(sizeof(_rc_values_previous) == sizeof(input_rc.values), "check sizeof(_rc_values_previous)");
 	}
 
 	perf_end(_loop_perf);
 }
 
-int
-RCUpdate::task_spawn(int argc, char *argv[])
+switch_pos_t RCUpdate::get_rc_sw3pos_position(uint8_t func, float on_th, float mid_th) const
+{
+	if (_rc.function[func] >= 0) {
+		const bool on_inv = (on_th < 0.f);
+		const bool mid_inv = (mid_th < 0.f);
+
+		const float value = 0.5f * _rc.channels[_rc.function[func]] + 0.5f;
+
+		if (on_inv ? value < on_th : value > on_th) {
+			return manual_control_switches_s::SWITCH_POS_ON;
+
+		} else if (mid_inv ? value < mid_th : value > mid_th) {
+			return manual_control_switches_s::SWITCH_POS_MIDDLE;
+
+		} else {
+			return manual_control_switches_s::SWITCH_POS_OFF;
+		}
+	}
+
+	return manual_control_switches_s::SWITCH_POS_NONE;
+}
+
+switch_pos_t RCUpdate::get_rc_sw2pos_position(uint8_t func, float on_th) const
+{
+	if (_rc.function[func] >= 0) {
+		const bool on_inv = (on_th < 0.f);
+
+		const float value = 0.5f * _rc.channels[_rc.function[func]] + 0.5f;
+
+		if (on_inv ? value < on_th : value > on_th) {
+			return manual_control_switches_s::SWITCH_POS_ON;
+
+		} else {
+			return manual_control_switches_s::SWITCH_POS_OFF;
+		}
+	}
+
+	return manual_control_switches_s::SWITCH_POS_NONE;
+}
+
+void RCUpdate::UpdateManualSwitches(const hrt_abstime &timestamp_sample)
+{
+	manual_control_switches_s switches{};
+	switches.timestamp_sample = timestamp_sample;
+
+	// check mode slot (RC_MAP_FLTMODE) or legacy mode switch (RC_MAP_MODE_SW), but not both
+	if (_param_rc_map_fltmode.get() > 0) {
```

Comment:
Filling `mode_slot`, `man_switch`, ... only under certain conditions relys on `MODE_SLOT_NONE` and `SWITCH_POS_NONE` staying zero for crrect initialization. This was not the case before.

Reply:
Do you mean somewhere else? Here the parameters `RC_MAP_FLTMODE` and `RC_MAP_MODE_SW` only assigned when > 0.

https://github.com/PX4/PX4-Autopilot/blob/d44e537084cdaf1a2ce001d6865f41b1013c4a9c/src/modules/rc_update/params.c#L1311

Reply:
Presumably you're talking about `manual_control_switches_s switches{}` just above? Yes I'm fine with zeroing being equivalent to SWITCH_POS_NONE. Do you see an issue?

Reply:
The potential issue is that the way a uORB::Subscription initializes you'll technically have an update right away with any existing data. It doesn't matter for most modules, but in the case of FlightTasks the subscriptions are being constructed on every flight task change.

That's why I added an initial gear_switch copy in the FlightTaskManualAltitude() constructor.

https://github.com/PX4/PX4-Autopilot/pull/16270/files#diff-292a0ed3849d2db6c5cd556e7e83aa5d93e78080c5ced1c6e8e6a2d9e9f6b37eR50-R52

Reply:
I simply wasn't aware that GEAR_KEEP was zero. Fixed.

FYI I'd like to see if we can get proper default values for msgs in the future.